### PR TITLE
Major improvements of Duckiebot interface

### PR DESCRIPTION
**Pull request by AleksandarPetrov**
**_TO BE TESTED AND MERGED TOGETHER WITH:_**
- dt-car-interface: https://github.com/duckietown/dt-car-interface/pull/2
- dt-core: https://github.com/duckietown/dt-core/pull/14
- Also related to: https://github.com/duckietown/dt-duckiebot-interface/pull/22

## Description

- All nodes in the Duckiebot interface are cleaned, rewritten where needed and documented
- Documentation framework updated and simplified, no explicit modulification of packages needed, a README with explicit instructions on building the documentation was also added
- Everything uses DTROS
- READMEs and .autogenerated.mds removed 
- `dagu_car` major updates: merged inverse and forward kinematics into a kinematics node, removed unused trimemer node, cleanup and optimization
- JoyMapperNode puts a timestamp on the messages it sends (necessary for the odometry to work)
- LEDEmitterNode rewritten and cleanedup
- rgb_led moved as a library of the led_emitter package
- traffic_light repaired and cleaned up
- various small fixes
- Cleanup of all __init__.py files
- Cleanup of readmes and .autogenerated.md files
- A lot of extra documentation
- All clear_params from the launch files removed (tnx to @AmaurX)
- Moved dagu and joy_mapper to the newly made car_interface

## Type of change

Please mark the relevant options with X:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is constrained only to one package, any interfaces to nodes from other packages are maintained
- [ ] This change is constrained only to packages in this repository, any interfaces to nodes from other repositories are maintained
- [X] This change required a documentation update
- [ ] This change is only a documentation update
- [X] This change restructures the repository (at level higher than package level)
- [ ] This change requires a change in the Duckiebook (if yes, please describe and, if possible, make a pull request for an edit in the book and link it here)

## Tests

Please describe all tests ran and their outcome. It is highly recommended that you at least do the tests listed by default, even if you only changed the documentation.

- [X] Made the Duckiebot move with keyboard control
- [X] Visually checked the published camera stream
- [X] Verified that the lane following demo works
- [X] Verified that the indefinite navigation demo works

## Style compliance
Please make sure that all your changes conform to the Duckietown (code style)[link] and code (documentation style)[link] guidelines:

- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I checked that this pull request is compliant with the Duckietown code style
- [X] I checked that this pull request is compliant with the Duckietown code documentation style

## Next steps
Please describe if you have any questions, comments, or doubts towards the Duckietown software guardians regarding implementation, code style, running tests, documentation, etc.

## FOR THE REVIEWERS
__IMPORTANT:__ DO NOT CHANGE THIS SECTION!

- [ ] Functionality (is the change necessary and is it implemented in the most appropriate way): @liampaull or @afdaniele
- [ ] Integration testing (verify the tests and run additional if necessary): @?
- [X] Code style compliance: @aleksandarpetrov
- [X] Documentation compliance: @aleksandarpetrov

_To the reviewers:_ If you deem necessary, request code edits, more tests, documentation update, or additional reviewing. __Do not review your own code!__

_PR Template, last edit 23 Aug 2019_

### CODE REVIEW

File: packages/traffic_light/src/traffic_light_node.py [hunk 1]
```diff
@@ -1,64 +1,66 @@
...
-        # Publishers
-        self.pub_pattern = rospy.Publisher("~custom_pattern",
-                                           LEDPattern,
-                                           queue_size=1)
-        # Initialize timer for direction change
+        # Start a timer that will regularly call a method that changes
+        # the direction that get green light
         self.traffic_cycle = rospy.Timer(rospy.Duration(cycle_duration),
                                          self.change_direction)
-        # Scale intensity of the LEDs & convert to right order
```
Comment (gibernas) [inline packages/traffic_light/src/traffic_light_node.py]: Are you dimming the lights somewhere else?

File: packages/led_emitter/config/led_emitter_node/LED_protocol.yaml [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
     red: [1,0,0]
     blue: [0,0,1]
-    white: [1,1,1]
     yellow: [1,1,0]
     purple: [1,0,1]
     cyan: [0,1,1]
-    black: [0,0,0]
+    pink: [1,0,0.5]
+    tl_red: [0,1,0]   # used for the traffic light LEDs as they are differently ordered
+    tl_green: [1,0,0] # used for the traffic light LEDs as they are differently ordered
```
Comment (gibernas) [inline packages/led_emitter/config/led_emitter_node/LED_protocol.yaml]: Isn´t it confusing to have multiple times the same color?
Also if we use more colors on TL we would have double the entries

File: packages/traffic_light/src/traffic_light_node.py (hunk 1 above)
Comment (AleksandarPetrov) [reply packages/traffic_light/src/traffic_light_node.py]: Yes, it uses the LEDEmitterNode and it handles the scaling. I have forgotten to remove this comment. Will handle it now.

File: packages/led_emitter/config/led_emitter_node/LED_protocol.yaml (hunk 2 above)
Comment (AleksandarPetrov) [reply packages/led_emitter/config/led_emitter_node/LED_protocol.yaml]: The problem comes from the different ordering of the RGB channels on a bot and a tower. Before, the traffic light node loaded this LED_protocol (which is not part of its configuration) and modified the colors in it. In my opinion it was dirty and non-modular solution. This is also not ideal, but I do not foresee additional colors on a traffic light (maybe yellow but nothing more).
Alternatively, we can have an additional parameter that controls the channel order, and on launch time, depending to whether we launch it on a bot or a tower, we use a different value for it. What do you think?

File: packages/duckiebot_interface/launch/all_drivers.launch [hunk 3]
```diff
@@ -15,8 +15,18 @@
...
             <arg name="veh" value="$(arg veh)"/>
         </include>
+
+        <!-- Adjust the ordering of the LED channels for the duckiebot and the traffic light  -->
+        <!-- as their hardware expects different order -->
+        <group if="$(eval arg('robot_type') == 'duckiebot')">
+            <param name="led_emitter_node/channel_order" value="RGB"/>
+        </group>
+        <group if="$(eval arg('robot_type') == 'traffic_light')">
+            <param name="led_emitter_node/channel_order" value="GRB"/>
```
[the remark this answers is not in the document]
Comment (AleksandarPetrov) [reply packages/duckiebot_interface/launch/all_drivers.launch]: @gibernas testing tomorrow as we made some changes today

File: docs/source/packages/traffic_light.rst [hunk 4]
```diff
@@ -1,4 +1,11 @@
 traffic\_light package
 ======================
 
-.. automodule:: traffic_light
+.. contents::
+
+The `traffic_light` package blablabla
```
[the remark this answers is not in the document]
Comment (AleksandarPetrov) [reply docs/source/packages/traffic_light.rst]: Yeah... feel free to volunteer yourself to fill these in, otherwise I will do them at some point

**Comment by AleksandarPetrov** [comment]
Tested. Merging. There might still be traffic-light issues

**Pull request merged**